src/libs/sys-util-rs/src/hooks.rs

```diff
@@ -0,0 +1,423 @@
+// Copyright (c) 2019-2021 Alibaba Cloud
+// Copyright (c) 2019-2021 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::collections::hash_map::DefaultHasher;
+use std::collections::HashMap;
+use std::ffi::OsString;
+use std::hash::{Hash, Hasher};
+use std::io::{self, Read, Result};
+use std::path::Path;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use lazy_static::lazy_static;
+use subprocess::{ExitStatus, Popen, PopenConfig, PopenError, Redirection};
+
+use crate::{eother, sl};
+
+const DEFAULT_HOOK_TIMEOUT: i32 = 10;
+
+/// Execution state of OCI hooks.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum HookState {
+    /// Hook is pending for executing/retry.
+    Pending,
+    /// Hook has been successfully executed.
+    Done,
+    /// Hook has been marked as ignore.
+    Ignored,
+}
+
+lazy_static! {
+    static ref EXEC_RESULT: Arc<Mutex<HashMap<u64, HookState>>> =
+        Arc::new(Mutex::new(HashMap::new()));
+}
+
+fn calculate_hash(hook: &oci::Hook) -> u64 {
```

Comment:
This whole approach looks dangerous to me.  There's no guarantee you won't have hash collisions AFAICT.  Why do you need to hash contents anyway?  Can't you just construct objects for each hook in the spec and keep track of those objects?

Reply:
If you absolutely must hash the hook, can't you use the [`Hash` trait](https://doc.rust-lang.org/std/hash/trait.Hash.html) to avoid the error-prone manual hashing code?